### PR TITLE
chore(snapshots): unify sparse and normal FS output paths

### PR DESCRIPTION
**Pull request by adowair**
This commit refactors the code paths that exercise normal and sparse
writing of restored content to the local filesystem. The goal is to
expose sparsefile.Copy() and iocopy.Copy() to be interchangeable,
thereby allowing us to wrap or transform them more easily in the future.

### CODE REVIEW

File: snapshot/restore/local_fs_output.go
Comment (adowair) [reply snapshot/restore/local_fs_output.go:377]: That's a good point, there's no need for allocation. I'll make this an internal `ioCopy()` instead.

File: snapshot/restore/local_fs_output.go
Comment (adowair) [reply snapshot/restore/local_fs_output.go:30]: Done: changed to `streamCopier` to avoid conflict with existing `copier` in package

File: snapshot/restore/local_fs_output.go
Comment (jkowalski) [reply snapshot/restore/local_fs_output.go:377]: one last nit: perhaps introduce `getStreamCopier()`, otherwise lgtm.

**Review by jkowalski — APPROVED** [review]

File: internal/sparsefile/sparsefile.go
Comment (adowair) [inline internal/sparsefile/sparsefile.go:19]: I realize now one thing missing here is we pass the entire buffer to `copyBuffer` (64K), not the intended disk block size (usually `512B`). This makes the current implementation of E2E test fail.

The next effect of this is that zero block detection/seeking will happen at that higher granularity, which is less than ideal. On the other hand, we can now enable sparse restores for Windows without much work, as we will no longer need to implement `GetDiskBlockSize()`–something that proved to be tricky for Windows.

@Shrekster we had a discussion earlier about this and I think we agreed this is an okay tradeoff. @jkowalski What do you think?

Comment (adowair) [inline internal/sparsefile/sparsefile.go:19]: I realize now one thing missing here is we pass the entire buffer to `copyBuffer` (64K), not the intended disk block size (usually `512B`). This makes the current implementation of E2E test fail.

The next effect of this is that zero block detection/seeking will happen at that higher granularity, which is less than ideal. On the other hand, we can now enable sparse restores for Windows without much work, as we will no longer need to implement `GetDiskBlockSize()`–something that proved to be tricky for Windows.

@Shrekster we had a discussion earlier about this and I think we agreed this is an okay tradeoff. @jkowalski What do you think?

Comment (adowair) [reply internal/sparsefile/sparsefile.go:19]: We can also just hardcode `512B` in...

Comment (adowair) [reply internal/sparsefile/sparsefile.go]: Resolved. Got around this by simply using a closure. Windows still not enabled, I'd like to work on that in a future PR =)

**Comment by adowair** [comment]
Ran the HTML E2E job locally (MacOS) and on a fork, and it seems to pass. The run fails here due to a timeout, which makes me think it is a flake.

File: snapshot/restore/local_fs_output.go
Comment (jkowalski) [inline snapshot/restore/local_fs_output.go:293]: nit: we can do this once per FilesystemOutput since it's not going to change. Perhaps call it right at the very beginning?

**Comment by jkowalski** [comment]
The flakes are known, we should fix them, but you don't worry about them for now. I have only minor tweaks, still LGTM, feel free to merge after those are addressed.

File: snapshot/restore/local_fs_output.go
Comment (adowair) [reply snapshot/restore/local_fs_output.go:293]: Done: moved `getStreamCopier()` to be public and called only once when first constructing `FilesystemOutput`.

File: snapshot/restore/local_fs_output.go
Comment (jkowalski) [inline snapshot/restore/local_fs_output.go:92]: this won't work, unfortunately. when serialized to JSON.

Since FilesystemOutput is passed verbatim from the UI, we need to pass `sparse` as boolean and resolve it later.

Comment (adowair) [reply snapshot/restore/local_fs_output.go:92]: Ah good point, I haven't thought of implementing this in UI, and should. Will do

Comment (adowair) [reply snapshot/restore/local_fs_output.go:92]: Reintroducing `Sparse` as an exported field, and made `Copier` private, to be assigned on the first call to `o.WriteFile()`.

Comment (adowair) [reply snapshot/restore/local_fs_output.go:92]: Correction, it looks like initializing copier upon the first call to `Writefile()` is not threadsafe, and has caused the race detector test to fail. Introducing `Init()` to replace `NewFilesystemOutput()` to do this assignment. Will call `Init()` for both server and CLI paths.

**Pull request merged**
